Add 10M MAC+PHY for the CH32V208

chmousset opened the pull request:
The CH32V208 has a MAC that's unique to itself, but relatively simple.
I was able to make the demo work on some custom HW of mine.

While this is functional, it's not the 'cleanest' implementation.
Feedback welcome.

This PR depends on https://github.com/ch32-rs/ch32-data/pull/16 for the registers definitions.
Once merged, the dependency yo ch3é-metapac should be updated here.


Note: This is incompatible with the MAC inside the CH32V30x.
While there is no support for the CH32V30x MAC, the basic infrastructure to include the correct ETH driver is present.
Also, the CH32V30x has a MAC that's very close to STM32's 'v1a' so [embassy-stm32](https://github.com/embassy-rs/embassy/tree/main/embassy-stm32/src/eth/v1) should be a good starting point for future work.

MDr164 commented:
For the CH32V307 mac there is now https://github.com/ch32-rs/ch32-hal/pull/157 to add support for it as well

chmousset committed:
Add 10M Ethernet MAC+PHY driver for CH32V208

Implements an embassy-net compatible driver for the built-in 10M Ethernet
MAC+PHY on CH32V208 microcontrollers. Uses embassy-net-driver-channel for
multi-buffer RX/TX packet queues.

- Add eth_10m driver (src/eth/v10m/mod.rs) with ISR-driven RX/TX/link handling
- Add GenericSMI PHY driver (src/eth/generic_smi.rs)
- Add eth module routing and traits (src/eth/mod.rs)
- Add UDP echo (eth.rs) and TCP echo (eth_echo.rs) examples for CH32V208
- Add embassy-net-driver-channel dependency

chmousset commented:
I finally had the time to fix and cleanup this code.
Both examples work on my little board.

@Codetector1374 it's ready for you to review / merge

aq1018 commented on src/eth/generic_smi.rs:
```diff
@@ -0,0 +1,78 @@
+impl GenericSMI {
+    /// Set the SMI polling interval.
+    #[cfg(feature = "time")]
+    pub fn set_poll_interval(&mut self, poll_interval: Duration) {
```
I think this is never used? There is no `time` feature. Maybe just delete?

chmousset replied:
good catch!
I must have forgotten to cleanup.

aq1018 reviewed (changes requested):
Just a small nit-pick. I think `set_poll_interval` is never used because it's gated behind a non-existent `time` feature.

Also, I merged in quite a few PRs. do you mind doing a rebase so CI can be run against latest dependencies?

chmousset committed:
[fix] removed unused set_poll_interval

Signed-off-by: Charles-Henri Mousset <ch.mousset@gmail.com>

chmousset commented:
@aq1018 rebase & cleanup done; I re-tested on the hardware I have and it still works.
